## Function of "Fixed Point Iteration" Method

In [78]:
import math
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [105]:
def fixed_point_iteration(func, x0, a ,b, tol=1e-5, max_iters=100):
    """
    Solve for a root of f(x) = 0 using fixed-point iteration method.

    Parameters:
    -----------
    func : function
        The fixed-point function. It must take a single argument, x, and return a single value.
    x0 : float
        Initial guess for the root.
    a, b : float
        Interval [a, b] containing the root.
    tol : float
        Tolerance for convergence.
    max_iters : int
        Maximum number of iterations.

    Returns:
    --------
    x : float
        Approximate root of f(x) = 0.
    i : int
        Number of iterations used.
    convergence_rate : float
        The rate of convergence.
    """

    # Initialize variables
    x = x0
    iters = 1
    x1 = func(x)

    # Initialize table
    table = pd.DataFrame(columns=['iteration', 'x','x1', 'f(x)'])

    # Iterate until convergence or max iterations reached
    while iters < max_iters:
        table = table.append({'iteration': iters, 'x': x,'x1': x1, 'f(x)': func(x)}, ignore_index=True)
        x1 = func(x)
        if abs(x1 - x) < tol:
            break
        if x1 < a or x1 > b:
            print("Out out interval error")
            break
        x = x1
        iters += 1

    return x,table

## Defining the main function

In [106]:
def f(x):
    return x**2 - 4*math.e**(-x)-3

In [108]:
root,table = fixed_point_iteration(f, 1, 1, 2, tol=1e-5, max_iters=100) #x0 = 1 interval [1,2]
print(table.head())

Out out interval error
   iteration    x        x1      f(x)
0        1.0  1.0 -3.471518 -3.471518


*	In first iteration x1 becomes -3.47 and it is out of interval it is not possible to solve even if we dont define the interval because the function diverges with initial value 1.